File: kernel/src/sched/cgroup_controllers.zig
// SPDX-License-Identifier: GPL-2.0 OR MIT
// Zxyphor Kernel - Cgroup v2 Controllers and Resource Management
// CPU, memory, IO, PID, RDMA, HugeTLB, cpuset, misc controllers
// PSI (Pressure Stall Information), delegation, threaded cgroups
// More advanced than Linux 2026 cgroup v2 architecture

const std = @import("std");

// ============================================================================
// Cgroup v2 Hierarchy
// ============================================================================

pub const CGROUP_MAX_DEPTH: u32 = 32;
pub const CGROUP_MAX_CHILDREN: u32 = 4096;

pub const CgroupType = enum(u8) {
    domain = 0,
    domain_threaded = 1,
    domain_invalid = 2,
    threaded = 3,
};

pub const CgroupSubsysId = enum(u8) {
    cpu = 0,
    memory = 1,
    io = 2,
    pids = 3,
    rdma = 4,
    hugetlb = 5,
    cpuset = 6,
    misc = 7,
    perf_event = 8,
    freezer = 9,
    // Zxyphor
    zxy_energy = 10,
    zxy_network = 11,
};

pub const NR_CGROUP_SUBSYS: u32 = 12;

pub const CgroupFlags = packed struct {
    populate_frozen: bool,
    frozen: bool,
    kill: bool,
    // Controller-specific
    cpu_enabled: bool,
    memory_enabled: bool,
    io_enabled: bool,
    pids_enabled: bool,
    rdma_enabled: bool,
    hugetlb_enabled: bool,
    cpuset_enabled: bool,
    misc_enabled: bool,
    perf_event_enabled: bool,
    freezer_enabled: bool,
    zxy_energy_enabled: bool,
    zxy_network_enabled: bool,
    _padding: u1 = 0,
};

pub const CgroupNode = struct {
    // Identity
    id: u64,
    name: [256]u8,
    name_len: u16,
    // Hierarchy
    parent: ?*CgroupNode,
    children: [CGROUP_MAX_CHILDREN]?*CgroupNode,
    nr_children: u32,
    depth: u32,
    // Type
    cgroup_type: CgroupType,
    flags: CgroupFlags,
    // Subtree control
    subtree_control: u16,
    // Processes
    nr_procs: u32,
    nr_threads: u32,
    nr_dying_descendants: u32,
    nr_populated_csets: u32,
    // Freezer
    frozen: bool,
    // Events
    events_populated: u64,
    events_frozen: u64,
    // Pressure
    psi: PsiGroup,
    // Controllers
    cpu_ctrl: CpuController,
    memory_ctrl: MemoryController,
    io_ctrl: IoController,
    pids_ctrl: PidsController,
    cpuset_ctrl: CpusetController,
    hugetlb_ctrl: HugeTlbController,
};

// ============================================================================
// CPU Controller
// ============================================================================

pub const CpuController = struct {
    // cpu.weight (1-10000, default 100)
    weight: u32,
    weight_nice: i32,
    // cpu.max (quota period)
    max_quota_us: i64,    // -1 = max
    max_period_us: u64,   // Default 100000 (100ms)
    // cpu.max.burst
    burst_us: u64,
    // Statistics
    usage_usec: u64,
    user_usec: u64,
    system_usec: u64,
    nr_periods: u64,
    nr_throttled: u64,
    throttled_usec: u64,
    nr_bursts: u64,
    burst_usec: u64,
    // Uclamp
    uclamp_min: u32,     // 0-1024
    uclamp_max: u32,     // 0-1024
    // EEVDF/CFS
    idle: bool,           // cpu.idle (deprioritize)
    // Pressure
    pressure_some: u64,
    pressure_full: u64,

    pub fn is_throttled(self: *const CpuController) bool {
        return self.max_quota_us > 0 and self.throttled_usec > 0;
    }

    pub fn throttle_ratio(self: *const CpuController) f64 {
        if (self.nr_periods == 0) return 0.0;
        return @as(f64, @floatFromInt(self.nr_throttled)) /
            @as(f64, @floatFromInt(self.nr_periods));
    }
};

// ============================================================================
// Memory Controller
// ============================================================================

pub const MemoryController = struct {
    // Limits
    min: u64,             // memory.min (hard protection)
    low: u64,             // memory.low (best-effort protection)
    high: u64,            // memory.high (throttle)
    max: u64,             // memory.max (hard limit)
    // Swap
    swap_max: u64,        // memory.swap.max
    swap_high: u64,       // memory.swap.high
    // zswap
    zswap_max: u64,       // memory.zswap.max
    zswap_writeback: bool, // memory.zswap.writeback
    // Current usage
    current: u64,         // memory.current
    swap_current: u64,    // memory.swap.current
    zswap_current: u64,   // memory.zswap.current
    // Peak
    peak: u64,            // memory.peak
    swap_peak: u64,       // memory.swap.peak
    // Statistics (memory.stat)
    anon: u64,
    file: u64,
    kernel: u64,
    kernel_stack: u64,
    pagetables: u64,
    sec_pagetables: u64,
    percpu: u64,
    sock: u64,
    vmalloc: u64,
    shmem: u64,
    zswap: u64,
    zswapped: u64,
    file_mapped: u64,
    file_dirty: u64,
    file_writeback: u64,
    swapcached: u64,
    anon_thp: u64,
    file_thp: u64,
    shmem_thp: u64,
    inactive_anon: u64,
    active_anon: u64,
    inactive_file: u64,
    active_file: u64,
    unevictable: u64,
    slab_reclaimable: u64,
    slab_unreclaimable: u64,
    pgfault: u64,
    pgmajfault: u64,
    workingset_refault_anon: u64,
    workingset_refault_file: u64,
    workingset_activate_anon: u64,
    workingset_activate_file: u64,
    workingset_restore_anon: u64,
    workingset_restore_file: u64,
    workingset_nodereclaim: u64,
    pgrefill: u64,
    pgscan: u64,
    pgsteal: u64,
    pgactivate: u64,
    pgdeactivate: u64,
    pglazyfree: u64,
    pglazyfreed: u64,
    thp_fault_alloc: u64,
    thp_collapse_alloc: u64,
    // Events (memory.events)
    events_low: u64,
    events_high: u64,
    events_max: u64,
    events_oom: u64,
    events_oom_kill: u64,
    events_oom_group_kill: u64,
    // NUMA stats
    numa_stat: [64]MemcgNuma,
    nr_numa_nodes: u32,
    // OOM
    oom_group: bool,
    // Reclaim
    reclaim_active: u64,   // memory.reclaim
    // Pressure
    pressure_some: u64,
    pressure_full: u64,
};

pub const MemcgNuma = struct {
    node_id: u32,
    anon: u64,
    file: u64,
    kernel_stack: u64,
    pagetables: u64,
    shmem: u64,
    file_mapped: u64,
    file_dirty: u64,
    file_writeback: u64,
    swapcached: u64,
    anon_thp: u64,
    file_thp: u64,
    shmem_thp: u64,
    inactive_anon: u64,
    active_anon: u64,
    inactive_file: u64,
    active_file: u64,
    unevictable: u64,
};

// ============================================================================
// IO Controller (io.max / io.weight)
// ============================================================================

pub const IoController = struct {
    // io.weight (1-10000, default 100)
    default_weight: u32,
    // Per-device configs
    devices: [64]IoDeviceConfig,
    nr_devices: u32,
    // io.stat (per-device statistics)
    stats: [64]IoStat,
    nr_stats: u32,
    // io.pressure
    pressure_some: u64,
    pressure_full: u64,
};

pub const IoDeviceConfig = struct {
    major: u32,
    minor: u32,
    // io.max settings
    rbps: u64,        // Read bytes per second (0=max)
    wbps: u64,        // Write bytes per second
    riops: u64,       // Read I/O operations per second
    wiops: u64,       // Write I/O operations per second
    // io.weight
    weight: u32,
    // io.latency
    target_latency_us: u64,
};

pub const IoStat = struct {
    major: u32,
    minor: u32,
    rbytes: u64,
    wbytes: u64,
    rios: u64,
    wios: u64,
    dbytes: u64,       // Discard
    dios: u64,
    cost_usage: u64,
    cost_wait: u64,
    cost_indebt: u64,
    cost_indelay: u64,
};

// ============================================================================
// PIDs Controller
// ============================================================================

pub const PidsController = struct {
    max: i64,           // pids.max (-1 = max)
    current: u64,       // pids.current
    peak: u64,          // pids.peak
    events_max: u64,    // pids.events max-exceeded count
};

// ============================================================================
// Cpuset Controller
// ============================================================================

pub const CpusetController = struct {
    // cpuset.cpus
    cpus: [16]u64,      // CPU bitmask (1024 CPUs max)
    // cpuset.cpus.effective
    cpus_effective: [16]u64,
    // cpuset.mems
    mems: [16]u64,      // NUMA node bitmask
    // cpuset.mems.effective
    mems_effective: [16]u64,
    // cpuset.cpus.partition
    partition: CpusetPartition,
    // cpuset.cpus.exclusive
    cpus_exclusive: [16]u64,
    cpus_exclusive_effective: [16]u64,
    // Flags
    cpu_exclusive: bool,
    mem_exclusive: bool,
    mem_hardwall: bool,
    memory_migrate: bool,
    sched_load_balance: bool,
    spread_page: bool,
    spread_slab: bool,
    memory_pressure_enabled: bool,
    // Stats
    memory_pressure: u64,

    pub fn cpu_count(self: *const CpusetController) u32 {
        var count: u32 = 0;
        for (self.cpus) |word| {
            count += @popCount(word);
        }
        return count;
    }

    pub fn mem_count(self: *const CpusetController) u32 {
        var count: u32 = 0;
        for (self.mems) |word| {
            count += @popCount(word);
        }
        return count;
    }
};

pub const CpusetPartition = enum(u8) {
    member = 0,
    root = 1,
    isolated = 2,
};

// ============================================================================
// HugeTLB Controller
// ============================================================================

pub const HugeTlbController = struct {
    // Per page-size limits
    sizes: [4]HugeTlbSizeConfig,
    nr_sizes: u32,
};

pub const HugeTlbSizeConfig = struct {
    page_size: u64,        // e.g., 2MB, 1GB
    max: u64,              // Max bytes
    current: u64,
    rsvd_max: u64,         // Reserved max
    rsvd_current: u64,
    events_max: u64,
};

// ============================================================================
// PSI (Pressure Stall Information)
// ============================================================================

pub const PsiResource = enum(u8) {
    cpu = 0,
    memory = 1,
    io = 2,
    irq = 3,
};

pub const PsiState = enum(u8) {
    some = 0,
    full = 1,
};

pub const PsiGroup = struct {
    // Averages (fixed-point, *100)
    avg10: [4][2]u64,    // [resource][some/full]
    avg60: [4][2]u64,
    avg300: [4][2]u64,
    total: [4][2]u64,    // Total stall time (us)

    pub fn get_avg10(self: *const PsiGroup, resource: PsiResource, state: PsiState) f64 {
        return @as(f64, @floatFromInt(self.avg10[@intFromEnum(resource)][@intFromEnum(state)])) / 100.0;
    }

    pub fn get_avg60(self: *const PsiGroup, resource: PsiResource, state: PsiState) f64 {
        return @as(f64, @floatFromInt(self.avg60[@intFromEnum(resource)][@intFromEnum(state)])) / 100.0;
    }
};

pub const PsiTrigger = struct {
    state: PsiState,
    threshold_us: u64,
    window_us: u64,
    // Growth since window start
    growth: u64,
    pending: bool,
    last_event_time: u64,
};

// ============================================================================
// Misc Controller
// ============================================================================

pub const MiscResource = enum(u8) {
    sev = 0,           // AMD SEV
    sev_es = 1,
    sev_snp = 2,
    tdx = 3,           // Intel TDX
    // Zxyphor
    zxy_accel = 4,
};

pub const MiscController = struct {
    resources: [8]MiscResourceConfig,
    nr_resources: u32,
};

pub const MiscResourceConfig = struct {
    res_type: MiscResource,
    max: u64,
    current: u64,
    events_max: u64,
    capacity: u64,      // System-wide capacity
};

// ============================================================================
// Cgroup BPF
// ============================================================================

pub const CgroupBpfAttachType = enum(u8) {
    ingress = 0,
    egress = 1,
    sock_create = 2,
    sock_ops = 3,
    device = 4,
    bind4 = 5,
    bind6 = 6,
    post_bind4 = 7,
    post_bind6 = 8,
    connect4 = 9,
    connect6 = 10,
    sendmsg4 = 11,
    sendmsg6 = 12,
    recvmsg4 = 13,
    recvmsg6 = 14,
    getsockopt = 15,
    setsockopt = 16,
    getpeername4 = 17,
    getpeername6 = 18,
    getsockname4 = 19,
    getsockname6 = 20,
    sysctl = 21,
    inet_ingress = 22,
    inet_egress = 23,
    inet4_lookup = 24,
    inet6_lookup = 25,
    lsm = 26,
};

// ============================================================================
// Cgroup v2 Delegation and Notifications
// ============================================================================

pub const CgroupDelegation = struct {
    // Delegated subtree root
    cgroup: ?*CgroupNode,
    // Delegatee
    delegate_pid: i32,
    // Which controllers are delegated
    delegated_controllers: u16,
    // File permissions
    delegate_files: u64,  // Bitmask of which files
};

pub const CgroupEvent = struct {
    event_type: CgroupEventType,
    cgroup_id: u64,
    timestamp: u64,
    data: u64,
};

pub const CgroupEventType = enum(u8) {
    populated = 0,
    not_populated = 1,
    frozen = 2,
    thawed = 3,
    oom = 4,
    memory_high = 5,
    memory_max = 6,
    pids_max = 7,
    io_latency = 8,
};

// ============================================================================
// Cgroup v2 Subsystem Manager
// ============================================================================

pub const CgroupSubsystem = struct {
    // Root cgroup
    root: ?*CgroupNode,
    // Stats
    nr_cgroups: u64,
    nr_dying_cgroups: u64,
    // Configuration
    default_cpu_weight: u32,
    default_io_weight: u32,
    // Unified hierarchy mount path
    mount_path: [256]u8,
    // PSI (system-wide)
    system_psi: PsiGroup,
    // Available controllers
    available_controllers: u16,
    // Delegation
    default_delegation: CgroupDelegation,
    // Events
    total_events: u64,
    // BPF
    nr_bpf_programs: u32,
    // Zxyphor
    zxy_auto_tuning: bool,
    zxy_energy_controller: bool,
    initialized: bool,

    pub fn pressure_cpu_some(self: *const CgroupSubsystem) f64 {
        return self.system_psi.get_avg10(.cpu, .some);
    }

    pub fn pressure_memory_some(self: *const CgroupSubsystem) f64 {
        return self.system_psi.get_avg10(.memory, .some);
    }

    pub fn pressure_io_some(self: *const CgroupSubsystem) f64 {
        return self.system_psi.get_avg10(.io, .some);
    }
};

File: rust/src/proc/cgroup_controllers.rs
// Zxyphor Kernel - Cgroup v2 Controllers Advanced,
// CPU Controller, Memory Controller, IO Controller,
// PID Controller, RDMA Controller, HugeTLB Controller,
// Misc Controller, Freezer, Cpuset
// More advanced than Linux 2026 cgroup subsystem

use core::fmt;

// ============================================================================
// Cgroup v2 Core Types
// ============================================================================

/// Cgroup controller type
#[repr(u16)]
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum CgroupController {
    Cpu = 0,
    Memory = 1,
    Io = 2,
    Pids = 3,
    Cpuset = 4,
    Rdma = 5,
    HugeTlb = 6,
    Misc = 7,
    Perf = 8,
    Freezer = 9,
    // Cgroup v1 only
    Devices = 10,
    NetCls = 11,
    NetPrio = 12,
    Blkio = 13,
    // Zxyphor native
    ZxyIsolation = 100,
    ZxyEnergy = 101,
    ZxyQos = 102,
}

/// Cgroup type
#[repr(u8)]
#[derive(Debug, Clone, Copy)]
pub enum CgroupType {
    Domain = 0,
    DomainThreaded = 1,
    Threaded = 2,
    DomainInvalid = 3,
}

/// Cgroup subtree control mask
#[repr(transparent)]
#[derive(Debug, Clone, Copy)]
pub struct CgroupSubtreeControl(pub u32);

impl CgroupSubtreeControl {
    pub const CPU: Self = Self(1 << 0);
    pub const MEMORY: Self = Self(1 << 1);
    pub const IO: Self = Self(1 << 2);
    pub const PIDS: Self = Self(1 << 3);
    pub const CPUSET: Self = Self(1 << 4);
    pub const RDMA: Self = Self(1 << 5);
    pub const HUGETLB: Self = Self(1 << 6);
    pub const MISC: Self = Self(1 << 7);
}

/// Cgroup stat
#[repr(C)]
#[derive(Debug, Clone)]
pub struct CgroupStat {
    pub nr_descendants: u32,
    pub nr_dying_descendants: u32,
}

/// Cgroup events
#[repr(C)]
#[derive(Debug, Clone)]
pub struct CgroupEvents {
    pub populated: bool,
    pub frozen: bool,
}

/// Cgroup freezer state
#[repr(u8)]
#[derive(Debug, Clone, Copy)]
pub enum CgroupFreezerState {
    Thawed = 0,
    Freezing = 1,
    Frozen = 2,
}

// ============================================================================
// CPU Controller
// ============================================================================

/// CPU controller configuration
#[repr(C)]
#[derive(Debug, Clone)]
pub struct CpuControllerConfig {
    pub weight: u32,               // [1, 10000] default 100
    pub weight_nice: i32,          // [-20, 19] mapped to weight
    pub max_quota_us: i64,         // max bandwidth quota (-1 for unlimited)
    pub max_period_us: u64,        // bandwidth period (default 100000)
    pub burst_us: u64,             // burst bandwidth
    pub uclamp_min: u32,           // utilization clamp min [0, 1024]
    pub uclamp_max: u32,           // utilization clamp max [0, 1024]
    pub idle: bool,                // idle hint
}

/// CPU controller stats
#[repr(C)]
#[derive(Debug, Clone)]
pub struct CpuControllerStats {
    pub usage_usec: u64,
    pub user_usec: u64,
    pub system_usec: u64,
    pub nr_periods: u64,
    pub nr_throttled: u64,
    pub throttled_usec: u64,
    pub nr_bursts: u64,
    pub burst_usec: u64,
    // PSI
    pub psi_some_total_us: u64,
    pub psi_full_total_us: u64,
    pub psi_some_avg10: f32,
    pub psi_some_avg60: f32,
    pub psi_some_avg300: f32,
    pub psi_full_avg10: f32,
    pub psi_full_avg60: f32,
    pub psi_full_avg300: f32,
}

// ============================================================================
// Memory Controller
// ============================================================================

/// Memory controller configuration
#[repr(C)]
#[derive(Debug, Clone)]
pub struct MemControllerConfig {
    pub min: i64,                  // memory.min (hard minimum, -1 for default)
    pub low: i64,                  // memory.low (best-effort minimum)
    pub high: i64,                 // memory.high (throttle above this)
    pub max: i64,                  // memory.max (hard limit, -1 for max)
    pub swap_high: i64,
    pub swap_max: i64,
    pub zswap_max: i64,
    pub zswap_writeback: bool,
    pub oom_group: bool,
}

/// Memory controller stats
#[repr(C)]
#[derive(Debug, Clone)]
pub struct MemControllerStats {
    pub anon: u64,
    pub file: u64,
    pub kernel: u64,
    pub kernel_stack: u64,
    pub pagetables: u64,
    pub sec_pagetables: u64,
    pub percpu: u64,
    pub sock: u64,
    pub vmalloc: u64,
    pub shmem: u64,
    pub zswap: u64,
    pub zswapped: u64,
    pub file_mapped: u64,
    pub file_dirty: u64,
    pub file_writeback: u64,
    pub swapcached: u64,
    pub anon_thp: u64,
    pub file_thp: u64,
    pub shmem_thp: u64,
    pub inactive_anon: u64,
    pub active_anon: u64,
    pub inactive_file: u64,
    pub active_file: u64,
    pub unevictable: u64,
    pub slab_reclaimable: u64,
    pub slab_unreclaimable: u64,
    // Events
    pub pgfault: u64,
    pub pgmajfault: u64,
    pub pgrefill: u64,
    pub pgscan: u64,
    pub pgsteal: u64,
    pub pgactivate: u64,
    pub pgdeactivate: u64,
    pub pglazyfree: u64,
    pub pglazyfreed: u64,
    pub thp_fault_alloc: u64,
    pub thp_collapse_alloc: u64,
    // Watermark
    pub workingset_refault_anon: u64,
    pub workingset_refault_file: u64,
    pub workingset_activate_anon: u64,
    pub workingset_activate_file: u64,
    pub workingset_restore_anon: u64,
    pub workingset_restore_file: u64,
    pub workingset_nodereclaim: u64,
    // NUMA
    pub numa_pages_migrated: u64,
    pub numa_pte_updates: u64,
    pub numa_hint_faults: u64,
    // PSI
    pub psi_some_total_us: u64,
    pub psi_full_total_us: u64,
}

/// Memory events
#[repr(C)]
#[derive(Debug, Clone)]
pub struct MemEvents {
    pub low: u64,
    pub high: u64,
    pub max: u64,
    pub oom: u64,
    pub oom_kill: u64,
    pub oom_group_kill: u64,
}

// ============================================================================
// IO Controller
// ============================================================================

/// IO controller weight configuration
#[repr(C)]
#[derive(Debug, Clone)]
pub struct IoWeightConfig {
    pub default_weight: u32,
    pub nr_device_weights: u32,
    pub device_weights: [IoDeviceWeight; 16],
}

#[repr(C)]
#[derive(Debug, Clone)]
pub struct IoDeviceWeight {
    pub major: u32,
    pub minor: u32,
    pub weight: u32,
}

/// IO controller max configuration
#[repr(C)]
#[derive(Debug, Clone)]
pub struct IoMaxConfig {
    pub major: u32,
    pub minor: u32,
    pub rbps: i64,
    pub wbps: i64,
    pub riops: i64,
    pub wiops: i64,
}

/// IO controller stats
#[repr(C)]
#[derive(Debug, Clone)]
pub struct IoControllerStats {
    pub major: u32,
    pub minor: u32,
    pub rbytes: u64,
    pub wbytes: u64,
    pub rios: u64,
    pub wios: u64,
    pub dbytes: u64,
    pub dios: u64,
    pub cost_usage: u64,
    pub cost_wait: u64,
    pub cost_indebt: u64,
    pub cost_indelay: u64,
}

/// IO latency target
#[repr(C)]
#[derive(Debug, Clone)]
pub struct IoLatencyTarget {
    pub major: u32,
    pub minor: u32,
    pub target_us: u64,
}

// ============================================================================
// PID Controller
// ============================================================================

/// PID controller configuration
#[repr(C)]
#[derive(Debug, Clone)]
pub struct PidControllerConfig {
    pub max: i64,     // -1 for max (no limit), otherwise the limit
}

/// PID controller stats
#[repr(C)]
#[derive(Debug, Clone)]
pub struct PidControllerStats {
    pub current: u64,
    pub events_max: u64,
}

// ============================================================================
// Cpuset Controller
// ============================================================================

/// Cpuset partition type
#[repr(u8)]
#[derive(Debug, Clone, Copy)]
pub enum CpusetPartition {
    Member = 0,
    Root = 1,
    Isolated = 2,
}

/// Cpuset controller configuration
#[repr(C)]
#[derive(Debug, Clone)]
pub struct CpusetConfig {
    pub cpus: [u64; 4],        // bitmask for up to 256 CPUs
    pub cpus_effective: [u64; 4],
    pub mems: [u64; 4],        // NUMA node bitmask
    pub mems_effective: [u64; 4],
    pub partition: CpusetPartition,
    pub cpu_exclusive: bool,
    pub mem_exclusive: bool,
    pub mem_hardwall: bool,
    pub memory_migrate: bool,
    pub sched_load_balance: bool,
    pub memory_pressure_enabled: bool,
    pub memory_spread_page: bool,
    pub memory_spread_slab: bool,
    pub sched_relax_domain_level: i32,
}

// ============================================================================
// RDMA Controller
// ============================================================================

/// RDMA resource limits
#[repr(C)]
#[derive(Debug, Clone)]
pub struct RdmaResourceLimits {
    pub hca_handle_max: u32,
    pub hca_object_max: u32,
}

/// RDMA resource usage
#[repr(C)]
#[derive(Debug, Clone)]
pub struct RdmaResourceUsage {
    pub hca_handle_current: u32,
    pub hca_object_current: u32,
}

// ============================================================================
// HugeTLB Controller
// ============================================================================

/// HugeTLB controller limits per size
#[repr(C)]
#[derive(Debug, Clone)]
pub struct HugeTlbLimits {
    pub page_size_kb: u64,
    pub max: i64,
    pub rsvd_max: i64,
    pub current: u64,
    pub rsvd_current: u64,
    pub events_max: u64,
}

// ============================================================================
// Misc Controller
// ============================================================================

/// Misc resource type
#[repr(u8)]
#[derive(Debug, Clone, Copy)]
pub enum MiscResourceType {
    SevPages = 0,
    SevEsPages = 1,
    TdxPages = 2,
}

/// Misc resource config
#[repr(C)]
#[derive(Debug, Clone)]
pub struct MiscResourceConfig {
    pub res_type: MiscResourceType,
    pub max: i64,
    pub current: u64,
    pub events_max: u64,
}

// ============================================================================
// Cgroup Pressure Stall Information (PSI)
// ============================================================================

/// PSI trigger configuration
#[repr(C)]
#[derive(Debug, Clone)]
pub struct CgroupPsiTrigger {
    pub resource: CgroupPsiResource,
    pub threshold_us: u64,
    pub window_us: u64,
    pub full: bool,
}

#[repr(u8)]
#[derive(Debug, Clone, Copy)]
pub enum CgroupPsiResource {
    Cpu = 0,
    Memory = 1,
    Io = 2,
}

// ============================================================================
// Subsystem Manager
// ============================================================================

#[repr(C)]
pub struct CgroupAdvancedSubsystem {
    pub v2_enabled: bool,
    pub unified_hierarchy: bool,
    pub nr_cgroups: u64,
    pub nr_controllers: u32,
    pub active_controllers: CgroupSubtreeControl,
    pub nr_frozen: u32,
    pub nr_psi_triggers: u32,
    pub default_cpu_weight: u32,
    pub default_io_weight: u32,
    pub initialized: bool,
}

impl CgroupAdvancedSubsystem {
    pub const fn new() -> Self {
        Self {
            v2_enabled: true,
            unified_hierarchy: true,
            nr_cgroups: 0,
            nr_controllers: 0,
            active_controllers: CgroupSubtreeControl(0),
            nr_frozen: 0,
            nr_psi_triggers: 0,
            default_cpu_weight: 100,
            default_io_weight: 100,
            initialized: false,
        }
    }
}
